Solution v0.1

panicua opened the pull request.

panicua commented:
Just deny it, I forgot that I don't need to create pull request, I also want to implement other unnescessary methods for the dict + need to ask something on QnA about it.

Oleksl888 commented on app/main.py:
```diff
@@ -1,2 +1,108 @@
+    def _hash_size_check_increase(self) -> None:
+        """Supporting func to check and increase hash_table size when needed"""
+
+        if self.__len__() == int(self._hash_size * self._resize_factor):
```
you already defined __len__ method therefor you can use syntactic sugar here

Oleksl888 commented on app/main.py:
```diff
@@ -1,2 +1,108 @@
+        try:
+            key_hash_position = hash(key) % self._hash_size
+        except TypeError as e:
+            raise TypeError(e)
```
there not much sense in catching this error and reraising without adding any extra logic to it.  If you call hash() on unacceptable type TypeError will be raised anyway

Oleksl888 commented on app/main.py:
```diff
@@ -1,2 +1,108 @@
+        self._hash_size_check_increase()
+        key_hash_position = self._find_key_position(key)
+
+        while self._hash_table[key_hash_position]:
+            if self._hash_table[key_hash_position][0] == key:
+                self._hash_table[key_hash_position][1] = value
+                return print(f"value with key '{key}' were rewritten")
+            key_hash_position = (key_hash_position + 1) % self._hash_size
```
You repeat this block of code on line 31, 50 so move it to a function to make your code DRY

Oleksl888 commented on app/main.py:
```diff
@@ -1,2 +1,108 @@
+                return print(f"value with key '{key}' were rewritten")
+            key_hash_position = (key_hash_position + 1) % self._hash_size
+
+        self._hash_table[key_hash_position] = [key, value]
```
> Also, don't forget to store (key, hash, value) as a node in the hash table. A hash table may be stored as a simple list of nodes.

As per task requirements you also need to store hash as well. Would make sense to make a Node data structure and add it to your hash table

Oleksl888 commented on app/main.py:
```diff
@@ -1,2 +1,108 @@
+        raise KeyError("Key is not found in the Dictionary")
+
+    def pop(self, key: Hashable) -> Any:
+        result = self.__getitem__(key)
```
use syntactic sugar here since you have these methods implemented:
e.g. del self[key]

Oleksl888 commented on app/main.py:
```diff
@@ -1,2 +1,108 @@
+        self._length: int = 0
+        self._hash_table: list = [None] * 8
+        self._hash_size: int = 8
+        self._resize_factor: float | int = 2 / 3
```
better use Fraction type here

Oleksl888 commented on app/main.py:
```diff
@@ -1,2 +1,108 @@
+                                              % self._hash_size)
+                temp_hash_table[item_key_hash_position] = item
+        self._hash_table = temp_hash_table
+        print(f"Dictionary size were increased "
```
imagine you work with dicts and they resize like every second, you will be flooded by these messages :)
Also a lot of prints actually will consume your system resources so its ok to use them while you test things but remove it when you open a PR.

panicua commented:
I've spend 5~ more hours today just to fix/update/refactor. 
---First hard part were to refactor it correctly so I can use one method ```_while_key_hash_occupied``` for getitem/setitem magic methods.
---Second hard part were to understand why my del magic method doesn't work as it should be, appeared that problem where hiden deep in how algorythm of "linear probing" work.
 
I'll try to explain if you want to read:
Imagine we have hash table and the spot that we want to write in is already taken, so we need to jump to the next spot (in our case to right, so index + 1) untill we find an empty spot and write your Node there. But imagine now you have to delete the item, but it's real spot not where it's hash says, it's actually to the right of it on some kind of length so we need to run from that entry point where our hash says and run to the right the whole list untill we make sure that no Node with this key exists (it's the worst case scenario O(n)). But in average it's still O(1).

Oleksl888 commented:
>But imagine now you have to delete the item, but it's real spot not where it's hash says, it's actually to the right of it on some kind of length so we need to run from that entry point where our hash says and run to the right the whole list untill we make sure that no Node with this key exists (it's the worst case scenario O(n)). But in average it's still O(1).

I didn't look into your code thoroughly yet but I think logic of traversing the hash table in case of collision should be the same no matter if you insert, retrieve or delete the item. The only place where you might get some wrong values is when you have to resize table. But again there should be no problem if you resize table, recalculate all new positions for all exisitng values and then add new node. On the opposite if you delete an item you might need to resize after you removed the element.

panicua commented:
> > But imagine now you have to delete the item, but it's real spot not where it's hash says, it's actually to the right of it on some kind of length so we need to run from that entry point where our hash says and run to the right the whole list untill we make sure that no Node with this key exists (it's the worst case scenario O(n)). But in average it's still O(1).
> 
> I didn't look into your code thoroughly yet but I think logic of traversing the hash table in case of collision should be the same no matter if you insert, retrieve or delete the item. The only place where you might get some wrong values is when you have to resize table. But again there should be no problem if you resize table, recalculate all new positions for all exisitng values and then add new node. On the opposite if you delete an item you might need to resize after you removed the element.


I did not implement resizign after del btw, It's gonna get too deep for this task I belive. It's not that I'm lazy, but.. :D

GPT:
However, Python does have a mechanism to resize dictionaries to a smaller size if a large number of items are deleted, but this happens gradually and through a process called "rehashing." Rehashing occurs when the dictionary's size drops below a certain threshold (usually around 1/3 of the current size), and it needs to be shrunk to save memory.

So, to answer your question, deleting all 6 items from the dictionary won't immediately reduce its hash table size. Python will keep the hash table at its current size for efficiency, and it may eventually resize it smaller through rehashing if many more items are deleted.

YuriiKindrat commented on app/main.py:
```diff
@@ -1,2 +1,119 @@
+
+        raise KeyError("Key is not found in the Dictionary")
+
+    def pop(self, key: Hashable) -> Any:
+        result = self[key]
+        del self[key]
+        return result
```
thats how pop method should work )
![image](https://github.com/mate-academy/py-dictionary/assets/103201010/9320ade5-1bc7-4f2a-9e2e-9d5e7a81178c)

panicua replied:
> thats how pop method should work ) ![image](https://private-user-images.githubusercontent.com/103201010/322375404-9320ade5-1bc7-4f2a-9e2e-9d5e7a81178c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTMxNjQwMjMsIm5iZiI6MTcxMzE2MzcyMywicGF0aCI6Ii8xMDMyMDEwMTAvMzIyMzc1NDA0LTkzMjBhZGU1LTFiYzctNGYyYS05ZTJlLTlkNWU3YTgxMTc4Yy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDE1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQxNVQwNjQ4NDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kODVjYmY4ODI2ZDAwZTFhMGI5ZWVmZTE0MmM1MjdkYzhmOTQ4YjNkYTA4OTFiNmZlNDUzNzA4MjM1ZjE1NWMzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.gt9k5plv5hMonMDXqKWnwAXCvC4wqomYXTahhBVCdE8)

you mean that I need to return None in my case instead of key error? so like ```result = self.get(key)``` instead of ```result = self[key]```?

Oleksl888 commented on app/main.py:
```diff
@@ -1,2 +1,132 @@
+            result = self[key]
+            del self[key]
+        except KeyError:
+            return default
```
Please check your logic - you should raise KeyError in case if no key and no default value

Oleksl888 commented on app/node.py:
```diff
@@ -0,0 +1,9 @@
+
+
+@dataclass
+class Node:
```
I'd say Node must be defined inside of a Dictionary class - because it is part of its implementation

Oleksl888 commented on tests/test_main.py:
```diff
@@ -6,18 +6,18 @@
-#     for key, value in items:
-#         del dictionary[key]
-#     print(len(dictionary))
+@pytest.mark.timeout(5)
```
Please revert changes to tests

Oleksl888 commented on app/main.py:
```diff
@@ -1,2 +1,132 @@
+        key_hash_position = self._find_key_position(key)
+        index_len = 0
+
+        while index_len <= self._hash_size:
```
So why not modify  `_while_key_hash_occupied` method and reuse it here to avoid repetition?

panicua replied:
Because if we use this method with "del", we cannot garantee that we will look for an item(Node) even if entry point ```table[key_hash_position] == None``` it will immediately exit the loop and give wrong return. It can happen due to collision if we have 3 Nodes with starting point at table[1] for example, they will be written in table[1], table[2], table[3] due to collision and imagine we need to delete first Node at table[1], after that we'll have None, table[2], table[3], now we need to delete table[2] or table[3] but their keys will have same address to table as table[1] that is already deleted, they look at the condition in ```_while_key_hash_occupied``` and see that their entry point (table[1]) is None so they don't even check other cells in table to make sure that needed item even exists.
![NVIDIA_Share_MLxlV766aa](https://github.com/mate-academy/py-dictionary/assets/13889031/5658eadc-0cc4-42c6-ad7c-c7074eebaf3f)

and this condition guarantees us that even if we don't guess exact entry point for our table we'll look at every cells in table at least once (in worst case scenario, if we don't have this key in our table or if every key had same entry point in our table and you had collision every time)
![NVIDIA_Share_50azuCeClz](https://github.com/mate-academy/py-dictionary/assets/13889031/772deaa2-948b-4a2e-bd37-6ab2d246125d)

panicua commented:
changes:
Changed methods: ```__setitem__```, ```__getitem__```, ```__delitem__```.
Added methods: ```_move_table_right``` - to jump + 1 right in hash table
```_find_key_in_table``` - an universal method that used in ```__getitem__```, ```__delitem__``` has differences only in return, not in condition.
Deleted method: ```_while_key_hash_occupied```.

panicua commented:
changes:
Changed methods: ```__setitem__```, ```__getitem__```, ```__delitem__```.
Added methods: ```_move_table_right``` - to jump + 1 right in hash table
```_find_key_in_table``` - an universal method that used in ```__getitem__```, ```__delitem__``` has differences only in return, not in condition.
Deleted method: ```_while_key_hash_occupied```.

Oleksl888 commented on app/main.py:
```diff
@@ -1,2 +1,110 @@
+        table = self._hash_table
+        for _ in range(self._hash_size):
+            if table[hash_key] is not None and table[hash_key].key == key:
+                if action == "get":
```
So here you could just return `hash_key` (btw `index` would be more appropriate name) to the function that called this method and then if its __getitem__ that you calling it from you do `table[hash_key].value`
if it's __delitem__ you do 
```
 table[hash_key] = None
                    self._length -= 1

Oleksl888 commented on app/main.py:
```diff
@@ -1,2 +1,110 @@
+from typing import Any, Hashable
+from fractions import Fraction
+from app.node import Node
```
blank line here :)

Oleksl888 reviewed (approved):
Good job!
